htmresearch/frameworks/nlp/classification_model.py

```diff
@@ -502,3 +335,74 @@ def _deSerializeExtraData(self, extraDataDir):
     """
     pass
 
+
+  def _inferDocumentDetailed(self, tokenList, sortResults=True):
+    """
+    Run inference on the model with this list of tokens and return classification
+    results, sampleIds and distances.  By default this routine will tokenize the
+    document and classify using these tokens. A reset is issued after inference.
+    Repeated sampleIds ARE removed from the results.
+
+    @param tokenList (str)     The list of tokens for inference
+    @param sortResults (bool) If true the list of sampleIds and distances
+                              will be sorted in order of increasing distances.
+
+    @return  (numpy array) An array of size numLabels. Position i contains
+                           the likelihood that this token belongs to the i'th
+                           category. An array containing all zeros implies no
+                           decision could be made.
+             (list)        A list of unique sampleIds
+             (numpy array) An array of distances from this document to each
+                           sampleId
+    """
+    # Default implementation, can be overridden
+
+    # For each token run inference on the token and accumulate sum of distances
+    # from this token to all other sampleIds.
+    lastTokenIndex = len(tokenList) - 1
+    categoryVotes = numpy.zeros(self.numLabels)
+    distancesForEachId = {}
+    classifier = self.getClassifier()
+
+    for i, token in enumerate(tokenList):
+      votes, idList, distances = self.inferToken(token,
+                                     reset=int(i == lastTokenIndex),
+                                     returnDetailedResults=True,
+                                     sortResults=False)
+
+      if votes.sum() > 0:
+        categoryVotes[votes.argmax()] += 1
+
+        # For each unique id, keep the minimum distance to this token
+        for j, sampleId in enumerate(idList):
```

Comment:
Is there ever a scenario where `votes` is all zeros but there are still IDs and distance values? What about unsupervised experiments (and Imbu) where `numLabels` = 0? Then these min distance calculations should not be w/in the current if block.